File: conformance/runner/go/replay_runner.go
package main

// Wire-replay runner for the Go SDK.
//
// Mode-gate: this entrypoint runs only when WIRE_REPLAY_DIR is set. The
// existing main.go handles mock-mode and remains untouched aside from a
// thin gate at the top of main() that delegates here.
//
// We import go/pkg/generated directly. The unexported mappers
// (projectFromGenerated, etc.) intentionally stay private; this runner
// consumes the same internal generated types oapi-codegen produces. Do
// not "fix" this import — it is a deliberate seam for canary purposes.
// Per the BC5-readiness plan §"Per-language additions" (PR 3): the Go
// runner uses `json.Unmarshal(bodyBytes, &generated.<Type>{})` for the
// typed-decode boundary and walks the same bytes parsed into `any` for
// extras detection.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"

	generated "github.com/basecamp/basecamp-sdk/go/pkg/generated"
)

const ReplaySchemaVersion = 1

// Decoders map operation_id -> decoder fn. Each decoder unmarshals
// bodyText into the generated response type produced by oapi-codegen
// (the same types the SDK uses internally) and returns the unmarshal
// error, or nil on success. Keep this table in sync with
// LIVE_OPERATIONS in conformance/runner/typescript/live-dispatch.ts —
// the coverage gate enforces parity.
var decoders = map[string]func(bodyText string) error{
	"ListProjects": func(bt string) error {
		var v generated.ListProjectsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetProject": func(bt string) error {
		var v generated.GetProjectResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetMyAssignments": func(bt string) error {
		var v generated.GetMyAssignmentsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetMyCompletedAssignments": func(bt string) error {
		var v generated.GetMyCompletedAssignmentsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetMyDueAssignments": func(bt string) error {
		var v generated.GetMyDueAssignmentsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetMyNotifications": func(bt string) error {
		var v generated.GetMyNotificationsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetMyProfile": func(bt string) error {
		var v generated.GetMyProfileResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"GetTodoset": func(bt string) error {
		var v generated.GetTodosetResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"ListTodolists": func(bt string) error {
		var v generated.ListTodolistsResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
	"ListTodos": func(bt string) error {
		var v generated.ListTodosResponseContent
		return json.Unmarshal([]byte(bt), &v)
	},
}

var safeNameRE = regexp.MustCompile(`(?i)[^a-z0-9_-]+`)

func safeName(s string) string {
	return safeNameRE.ReplaceAllString(s, "_")
}

// ReplayPage mirrors the per-page output schema documented in the
// BC5-readiness plan §"Snapshot output". Slices are emitted as `[]` (not
// null) in the JSON output for downstream comparator stability — see
// emptyIfNil before serialization.
type ReplayPage struct {
	Decoded         bool     `json:"decoded"`
	DecodeError     *string  `json:"decode_error"`
	MissingRequired []string `json:"missing_required"`
	ExtrasSeen      []string `json:"extras_seen"`
}

type ReplayResult struct {
	SchemaVersion int          `json:"schema_version"`
	Operation     string       `json:"operation"`
	Pages         []ReplayPage `json:"pages"`
}

type fixtureTest struct {
	Mode      string `json:"mode"`
	Name      string `json:"name"`
	Operation string `json:"operation"`
}

type wirePage struct {
	Status   int               `json:"status"`
	Headers  map[string]string `json:"headers"`
	Body     any               `json:"body"`
	BodyText *string           `json:"bodyText"`
	URL      string            `json:"url"`
}

type wireSnapshot struct {
	Operation  string     `json:"operation"`
	Pages      []wirePage `json:"pages"`
	PagesCount int        `json:"pages_count"`
}

type ReplayRunner struct {
	replayDir   string
	backend     string
	fixturePath string
	walker      *SchemaWalker
	fixture     []fixtureTest
}

func NewReplayRunner(replayDir, backend, fixturePath, openapiPath string) (*ReplayRunner, error) {
	walker, err := NewSchemaWalker(openapiPath)
	if err != nil {
		return nil, err
	}
	raw, err := os.ReadFile(fixturePath)
	if err != nil {
		return nil, err
	}
	var all []fixtureTest
	if err := json.Unmarshal(raw, &all); err != nil {
		return nil, err
	}
	live := make([]fixtureTest, 0, len(all))
	for _, t := range all {
		if t.Mode == "live" {
			live = append(live, t)
		}
	}
	return &ReplayRunner{replayDir, backend, fixturePath, walker, live}, nil
}

// coverageGate enforces the three startup checks documented in PR 3
// §"Coverage gate":
//  1. every fixture op has a decoder
//  2. every fixture op has a snapshot file
//  3. every snapshot's operation field is in the fixture
//
// Any failure returns a non-empty list of human-readable messages; the
// caller prints them and exits non-zero.
func (r *ReplayRunner) coverageGate() []string {
	var msgs []string
	fixtureOps := map[string]bool{}
	for _, t := range r.fixture {
		fixtureOps[t.Operation] = true
	}

	// 1. Decoder coverage
	var missing []string
	for op := range fixtureOps {
		if _, ok := decoders[op]; !ok {
			missing = append(missing, op)
		}
	}
	if len(missing) > 0 {
		sort.Strings(missing)
		msgs = append(msgs, fmt.Sprintf(
			"Go replay runner missing decoders for: %v. Add to decoders map in replay_runner.go.",
			missing,
		))
	}

	// 2. Snapshot completeness
	wireDir := filepath.Join(r.replayDir, r.backend, "wire")
	for _, t := range r.fixture {
		f := filepath.Join(wireDir, safeName(t.Name)+".json")
		if _, err := os.Stat(f); err != nil {
			msgs = append(msgs, fmt.Sprintf(
				"Snapshot missing for operation %s (test %q); expected at %s. Re-run TS live capture or check skip status.",
				t.Operation, t.Name, f,
			))
		}
	}

	// 3. Snapshot recognition
	if entries, err := os.ReadDir(wireDir); err == nil {
		for _, e := range entries {
			if filepath.Ext(e.Name()) != ".json" {
				continue
			}
			data, err := os.ReadFile(filepath.Join(wireDir, e.Name()))
			if err != nil {
				continue
			}
			var snap struct {
				Operation string `json:"operation"`
			}
			if err := json.Unmarshal(data, &snap); err != nil {
				continue
			}
			if snap.Operation == "" {
				msgs = append(msgs, fmt.Sprintf(
					"Snapshot %s is missing the top-level `operation` field. Re-run the TS live canary; pre-PR3 snapshots are no longer supported.",
					e.Name(),
				))
				continue
			}
			if !fixtureOps[snap.Operation] {
				msgs = append(msgs, fmt.Sprintf(
					"Unknown operation %q in snapshot %s; TS dispatch table appears to have drifted from live-my-surface.json.",
					snap.Operation, e.Name(),
				))
			}
		}
	}

	return msgs
}

func (r *ReplayRunner) Run() int {
	if msgs := r.coverageGate(); len(msgs) > 0 {
		for _, m := range msgs {
			fmt.Fprintln(os.Stderr, m)
		}
		return 1
	}

	outDir := filepath.Join(r.replayDir, r.backend, "decode", "go")
	if err := os.MkdirAll(outDir, 0o755); err != nil {
		fmt.Fprintln(os.Stderr, err)
		return 1
	}

	failures := 0
	for _, t := range r.fixture {
		snap, err := r.readSnapshot(t.Name)
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			failures++
			continue
		}
		result := r.decodeSnapshot(snap)
		out, err := json.MarshalIndent(result, "", "  ")
		if err != nil {
			fmt.Fprintln(os.Stderr, err)
			failures++
			continue
		}
		outPath := filepath.Join(outDir, safeName(t.Name)+".json")
		if err := os.WriteFile(outPath, out, 0o644); err != nil {
			fmt.Fprintln(os.Stderr, err)
			failures++
			continue
		}
		for _, p := range result.Pages {
			if !p.Decoded || len(p.MissingRequired) > 0 {
				failures++
				break
			}
		}
	}

	if failures > 0 {
		return 1
	}
	return 0
}

func (r *ReplayRunner) readSnapshot(testName string) (*wireSnapshot, error) {
	path := filepath.Join(r.replayDir, r.backend, "wire", safeName(testName)+".json")
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, err
	}
	var snap wireSnapshot
	if err := json.Unmarshal(raw, &snap); err != nil {
		return nil, err
	}
	return &snap, nil
}

func (r *ReplayRunner) decodeSnapshot(snap *wireSnapshot) ReplayResult {
	decoder := decoders[snap.Operation]
	schema := r.walker.FindResponseSchema(snap.Operation)

	pages := make([]ReplayPage, 0, len(snap.Pages))
	for _, page := range snap.Pages {
		var bodyText string
		switch {
		case page.BodyText != nil:
			bodyText = *page.BodyText
		case page.Body != nil:
			if b, err := json.Marshal(page.Body); err == nil {
				bodyText = string(b)
			}
		}

		rp := ReplayPage{Decoded: false}
		if err := decoder(bodyText); err == nil {
			rp.Decoded = true
		} else {
			s := err.Error()
			rp.DecodeError = &s
		}

		if schema != nil {
			// Per the TS validator: walk against parsed JSON, not the
			// SDK-decoded structure (decoders may drop unknown fields).
			// A page whose body is not parseable JSON gets empty arrays
			// here — the decode_error above already captures that failure.
			var body any
			if json.Unmarshal([]byte(bodyText), &body) == nil {
				rp.MissingRequired = r.walker.MissingRequired(body, schema)
				rp.ExtrasSeen = r.walker.ExtrasSeen(body, schema)
			}
		}
		// Emit `[]` rather than `null` for stable cross-language
		// comparator output (PR 4 jq diffs would otherwise false-fire).
		if rp.MissingRequired == nil {
			rp.MissingRequired = []string{}
		}
		if rp.ExtrasSeen == nil {
			rp.ExtrasSeen = []string{}
		}
		pages = append(pages, rp)
	}

	return ReplayResult{
		SchemaVersion: ReplaySchemaVersion,
		Operation:     snap.Operation,
		Pages:         pages,
	}
}

File: kotlin/conformance/src/main/kotlin/com/basecamp/sdk/conformance/ReplayRunner.kt
package com.basecamp.sdk.conformance

import com.basecamp.sdk.generated.models.Person
import com.basecamp.sdk.generated.models.Project
import com.basecamp.sdk.generated.models.Todo
import com.basecamp.sdk.generated.models.Todolist
import com.basecamp.sdk.generated.models.Todoset
import kotlinx.serialization.Serializable
import kotlinx.serialization.SerializationException
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import java.io.IOException
import java.nio.file.Path
import java.nio.file.Paths
import kotlin.system.exitProcess

/**
 * Wire-replay runner for the Kotlin SDK conformance suite.
 *
 * Reads canonical wire snapshots written by the TS live runner (see
 * `conformance/runner/typescript/live-runner.test.ts`), decodes each page
 * through the Kotlin SDK's deserialization boundary, walks the raw JSON for
 * required-field and extras detection, and persists per-test decode-result
 * snapshots at `<WIRE_REPLAY_DIR>/<BACKEND>/decode/kotlin/<safe>.json`.
 *
 * Mode-gate: invoked only when `WIRE_REPLAY_DIR` is set (Makefile target
 * `conformance-kotlin-replay`). The existing mock runner (`Main.kt`) handles
 * the unset case and is unaffected by anything in this file.
 *
 * Decode boundary
 * ---------------
 * Where the Kotlin SDK has a typed model for the response payload (`Project`,
 * `Person`, `Todo`, `Todolist`, `Todoset`), the decoder routes through that
 * type — exactly what the generated service method does at runtime. The four
 * `My*` operations (assignments, completed assignments, due assignments,
 * notifications) currently return `JsonElement` from the SDK because
 * `MyAssignment` and `Notification` lack typed Kotlin models; for those we
 * decode as `JsonElement`, which mirrors the SDK's actual surface.
 *
 * The `Json` instance below intentionally mirrors mock-mode (`ignoreUnknownKeys
 * = true`, `coerceInputValues = false`) — additive BC5 fields must NOT decode
 * as errors here. This is a forward-compat canary, not a strictness audit.
 * Type mismatches and missing required (non-nullable) fields still throw and
 * surface as `decode_error`. Extras detection comes from the schema walker on
 * the parsed `JsonElement`, not from the decoder.
 */
private val replayJson = Json {
    ignoreUnknownKeys = true
    coerceInputValues = false
    isLenient = false
}

/** Walker uses an unconfigured Json — only parses the body for tree access. */
private val parserJson = Json

const val REPLAY_SCHEMA_VERSION = 1

/**
 * Decoders per operation. Each takes the raw bodyText for one page and
 * either completes normally (decode succeeded) or throws (decode failed).
 *
 * For `JsonElement`-returning SDK methods, we decode as `JsonElement` —
 * that's the SDK's actual decode boundary today. When the SDK adds typed
 * models for those operations, switch the decoder to the typed model.
 */
private val decoders: Map<String, (String) -> Unit> = mapOf(
    "ListProjects" to { bt -> replayJson.decodeFromString(ListSerializer(Project.serializer()), bt) },
    "GetProject" to { bt -> replayJson.decodeFromString(Project.serializer(), bt) },
    "GetMyAssignments" to { bt -> replayJson.decodeFromString(JsonElement.serializer(), bt) },
    "GetMyCompletedAssignments" to { bt -> replayJson.decodeFromString(JsonElement.serializer(), bt) },
    "GetMyDueAssignments" to { bt -> replayJson.decodeFromString(JsonElement.serializer(), bt) },
    "GetMyNotifications" to { bt -> replayJson.decodeFromString(JsonElement.serializer(), bt) },
    "GetMyProfile" to { bt -> replayJson.decodeFromString(Person.serializer(), bt) },
    "GetTodoset" to { bt -> replayJson.decodeFromString(Todoset.serializer(), bt) },
    "ListTodolists" to { bt -> replayJson.decodeFromString(ListSerializer(Todolist.serializer()), bt) },
    "ListTodos" to { bt -> replayJson.decodeFromString(ListSerializer(Todo.serializer()), bt) },
)

private val safeNameRegex = Regex("[^a-z0-9_-]+", RegexOption.IGNORE_CASE)
private fun safeName(s: String): String = safeNameRegex.replace(s, "_")

@Serializable
data class ReplayPageResult(
    val decoded: Boolean,
    val decode_error: String?,
    val missing_required: List<String>,
    val extras_seen: List<String>,
)

@Serializable
data class ReplayResult(
    val schema_version: Int,
    val operation: String,
    val pages: List<ReplayPageResult>,
)

class ReplayRunner(
    private val replayDir: Path,
    private val backend: String,
    fixturePath: Path,
    openapiPath: Path,
) {
    private val walker = SchemaWalker(openapiPath.toString())
    private val fixture: List<JsonObject> = run {
        val all = parserJson.parseToJsonElement(File(fixturePath.toString()).readText()).jsonArray
        all.mapNotNull { el ->
            val obj = el as? JsonObject ?: return@mapNotNull null
            obj.takeIf { it["mode"]?.jsonPrimitive?.contentOrNull == "live" }
        }
    }

    fun coverageGate(): List<String> {
        val msgs = mutableListOf<String>()
        val fixtureOps = fixture.mapNotNull { it["operation"]?.jsonPrimitive?.contentOrNull }
            .distinct().sorted()

        // 1. Decoder coverage — every fixture operation must have a decoder.
        val missing = fixtureOps.filterNot { decoders.containsKey(it) }
        if (missing.isNotEmpty()) {
            msgs += "Kotlin replay runner missing decoders for: ${missing.joinToString(", ")}. " +
                "Add to decoders in ReplayRunner.kt."
        }

        // 2. Snapshot completeness — every fixture op needs a wire file.
        val wireDir = replayDir.resolve(backend).resolve("wire").toFile()
        for (t in fixture) {
            val name = t["name"]?.jsonPrimitive?.contentOrNull ?: continue
            val op = t["operation"]?.jsonPrimitive?.contentOrNull ?: continue
            val f = File(wireDir, "${safeName(name)}.json")
            if (!f.exists()) {
                msgs += "Snapshot missing for operation $op (test \"$name\"); expected at " +
                    "${f.path}. Re-run TS live capture or check skip status."
            }
        }

        // 3. Snapshot recognition — every captured snapshot's operation must
        //    be in the shared fixture (catches TS-side dispatch drift).
        if (wireDir.exists()) {
            wireDir.listFiles { f -> f.extension == "json" }?.forEach { f ->
                val text = try {
                    f.readText()
                } catch (e: IOException) {
                    msgs += "Snapshot ${f.name} could not be read: ${e::class.simpleName}: ${e.message}."
                    return@forEach
                }
                val parsed = try {
                    parserJson.parseToJsonElement(text)
                } catch (e: SerializationException) {
                    msgs += "Snapshot ${f.name} is not valid JSON: ${e.message}."
                    return@forEach
                }
                val snap = parsed as? JsonObject
                if (snap == null) {
                    msgs += "Snapshot ${f.name} top-level JSON is not an object; expected the wire-snapshot envelope."
                    return@forEach
                }
                // Defensive: `.jsonPrimitive` on JsonNull/JsonObject/JsonArray
                // throws, which would crash the gate. Cast first so a malformed
                // `operation` value emits the gate message instead.
                val op = (snap["operation"] as? JsonPrimitive)?.contentOrNull
                if (op == null) {
                    msgs += "Snapshot ${f.name} is missing the top-level `operation` field. " +
                        "Re-run the TS live canary; pre-PR3 snapshots are no longer supported."
                    return@forEach
                }
                if (op !in fixtureOps) {
                    msgs += "Unknown operation \"$op\" in snapshot ${f.name}; TS dispatch " +
                        "table appears to have drifted from live-my-surface.json."
                }
            }
        }
        return msgs
    }

    fun run(): Int {
        val msgs = coverageGate()
        if (msgs.isNotEmpty()) {
            msgs.forEach { System.err.println(it) }
            return 1
        }

        val outDir = replayDir.resolve(backend).resolve("decode").resolve("kotlin").toFile()
        outDir.mkdirs()

        val pretty = Json { prettyPrint = true }
        var failures = 0
        for (t in fixture) {
            val name = t["name"]!!.jsonPrimitive.content
            val snap = readSnapshot(name)
            val result = decodeSnapshot(snap)
            File(outDir, "${safeName(name)}.json").writeText(
                pretty.encodeToString(ReplayResult.serializer(), result)
            )
            if (result.pages.any { !it.decoded || it.missing_required.isNotEmpty() }) failures++
        }
        return if (failures == 0) 0 else 1
    }

    private fun readSnapshot(testName: String): JsonObject {
        val path = replayDir.resolve(backend).resolve("wire")
            .resolve("${safeName(testName)}.json").toFile()
        return parserJson.parseToJsonElement(path.readText()).jsonObject
    }

    private fun decodeSnapshot(snap: JsonObject): ReplayResult {
        val operation = snap["operation"]!!.jsonPrimitive.content
        val decoder = decoders[operation]!!
        val schema = walker.findResponseSchema(operation)

        val pages = snap["pages"]!!.jsonArray.map { p ->
            val page = p.jsonObject
            // Prefer bodyText (raw); fall back to serializing `body` if absent.
            val bodyText = page["bodyText"]?.jsonPrimitive?.contentOrNull
                ?: page["body"]?.let { parserJson.encodeToString(JsonElement.serializer(), it) }
                ?: ""

            var decoded = false
            var decodeError: String? = null
            try {
                decoder(bodyText)
                decoded = true
            } catch (e: Exception) {
                // Catch Exception, not Throwable: fatal JVM errors (OOM,
                // StackOverflow, LinkageError) shouldn't be silently demoted
                // to a decode-failure record.
                decodeError = "${e::class.simpleName}: ${e.message}"
            }

            var missing: List<String> = emptyList()
            var extras: List<String> = emptyList()
            if (schema != null && bodyText.isNotEmpty()) {
                try {
                    val body = parserJson.parseToJsonElement(bodyText)
                    missing = walker.missingRequired(body, schema)
                    extras = walker.extrasSeen(body, schema)
                } catch (_: Exception) {
                    // Body is not parseable JSON; decode_error above already
                    // records the failure. Leave walker output empty.
                }
            }

            ReplayPageResult(decoded, decodeError, missing, extras)
        }
        return ReplayResult(REPLAY_SCHEMA_VERSION, operation, pages)
    }
}

fun main() {
    // Treat empty/blank as missing — Gradle and shells can pass empty strings
    // through, and those would otherwise slip past a `?: null` guard and let
    // the runner read/write relative to CWD or `<dir>/<empty>/...`.
    val replayDir = System.getenv("WIRE_REPLAY_DIR")?.takeUnless { it.isBlank() }
        ?: run { System.err.println("WIRE_REPLAY_DIR is required"); exitProcess(1) }
    val backend = System.getenv("BASECAMP_BACKEND")?.takeUnless { it.isBlank() }
        ?: run { System.err.println("BASECAMP_BACKEND is required"); exitProcess(1) }

    // Resolve repo paths relative to the kotlin/ root. The Gradle `runReplay`
    // task sets workingDir = rootProject.projectDir (i.e. the repo's `kotlin/`
    // directory), so repoRoot is the parent.
    val kotlinRoot = Paths.get("").toAbsolutePath()
    val repoRoot = kotlinRoot.parent
        ?: error("Cannot resolve repo root from CWD=$kotlinRoot")
    val fixturePath = repoRoot.resolve("conformance/tests/live-my-surface.json")
    val openapiPath = repoRoot.resolve("openapi.json")

    val runner = ReplayRunner(Paths.get(replayDir), backend, fixturePath, openapiPath)
    exitProcess(runner.run())
}
